File: mapping.py
import csv
import numpy as np
from jinja2 import Environment, FileSystemLoader

initial_value = float(input("give the initial value:"))
with open("spring.csv") as f:
    reader = csv.reader(f)
    l = np.array(list(reader), dtype=np.float64)

l[:, 0] -= initial_value
l[:, 0] = -l[:, 0]
l[:, 0] *= 10
i = np.argsort(l[:, 1])
l = l[i]
datas = []
for entry in zip(l[:-1], l[1:]):
    (y1, x1), (y2, x2) = entry
    x1, y1 = int(x1), float(y1)
    x2, y2 = int(x2), float(y2)
    k = (y2 - y1) / (x2 - x1)
    b = y1 - k * x1
    datas.append({"x1": x1, "x2": x2, "k": k, "b": b})
env = Environment(loader=FileSystemLoader("."))
template = env.get_template('mapping_c.jinja2')
mapping = template.render(datas=datas)
with open("mapping.c", "w") as fh:
    fh.write(mapping)

File: mapping.c

    
        if (proximityData < -33) {
        input = 0.482 * proximityData + 5.911;
        }
    

    
        else if (proximityData < 12) {
        input = 0.222 * proximityData + -2.667;
        }
    

    
        else if (proximityData < 58) {
        input = 0.054 * proximityData + -0.652;
        }
    

    
        else if (proximityData < 127) {
        input = 0.13 * proximityData + -5.065;
        }
    

    
        else if (proximityData < 223) {
        input = 0.13 * proximityData + -5.036;
        }
    

    
        else if (proximityData < 320) {
        input = 0.062 * proximityData + 10.206;
        }
    

    
        else if (proximityData < 548) {
        input = 0.022 * proximityData + 22.982;
        }
    

    
        else if (proximityData < 845) {
        input = 0.03 * proximityData + 18.394;
        }
    

    
        else if (proximityData < 1197) {
        input = 0.033 * proximityData + 16.393;
        }
    

    
        else if (proximityData < 1803) {
        input = 0.017 * proximityData + 34.76;
        }
    

    
        else if (proximityData < 2606) {
        input = 0.008 * proximityData + 51.405;
        }
    

    
        else if (proximityData < 4115) {
        input = 0.004 * proximityData + 61.275;
        }
    

    
        else if (proximityData < 6480) {
        input = 0.004 * proximityData + 61.6;
        }
    

    
        else if (proximityData < 10446) {
        input = 0.002 * proximityData + 79.197;
        }
    

    
        else {
        input = 0.001 * proximityData + 80.588;
        }
    
